File: clients/FSharp/Model/EntityProperties.fs
#nowarn "0058"
namespace Aicup2020.Model
type EntityProperties = {
    Size: int;
    BuildScore: int;
    DestroyScore: int;
    CanMove: bool;
    PopulationProvide: int;
    PopulationUse: int;
    MaxHealth: int;
    Cost: int;
    SightRange: int;
    ResourcePerHealth: int;
    Build: option<BuildProperties>;
    Attack: option<AttackProperties>;
    Repair: option<RepairProperties>;
    } with
    member this.writeTo(writer: System.IO.BinaryWriter) =
        writer.Write this.Size
        writer.Write this.BuildScore
        writer.Write this.DestroyScore
        writer.Write this.CanMove
        writer.Write this.PopulationProvide
        writer.Write this.PopulationUse
        writer.Write this.MaxHealth
        writer.Write this.Cost
        writer.Write this.SightRange
        writer.Write this.ResourcePerHealth
        match this.Build with
            | Some value ->
                writer.Write true
                value.writeTo writer
            | None -> writer.Write false
        match this.Attack with
            | Some value ->
                writer.Write true
                value.writeTo writer
            | None -> writer.Write false
        match this.Repair with
            | Some value ->
                writer.Write true
                value.writeTo writer
            | None -> writer.Write false
    static member readFrom(reader: System.IO.BinaryReader) = {
        Size = reader.ReadInt32()
        BuildScore = reader.ReadInt32()
        DestroyScore = reader.ReadInt32()
        CanMove = reader.ReadBoolean()
        PopulationProvide = reader.ReadInt32()
        PopulationUse = reader.ReadInt32()
        MaxHealth = reader.ReadInt32()
        Cost = reader.ReadInt32()
        SightRange = reader.ReadInt32()
        ResourcePerHealth = reader.ReadInt32()
        Build = match reader.ReadBoolean() with
            | true ->
                Some(
                    BuildProperties.readFrom reader
                    )
            | false -> None
        Attack = match reader.ReadBoolean() with
            | true ->
                Some(
                    AttackProperties.readFrom reader
                    )
            | false -> None
        Repair = match reader.ReadBoolean() with
            | true ->
                Some(
                    RepairProperties.readFrom reader
                    )
            | false -> None
    }

File: clients/CSharp/Model/EntityProperties.cs
namespace Aicup2020.Model
{
    public struct EntityProperties
    {
        public int Size { get; set; }
        public int BuildScore { get; set; }
        public int DestroyScore { get; set; }
        public bool CanMove { get; set; }
        public int PopulationProvide { get; set; }
        public int PopulationUse { get; set; }
        public int MaxHealth { get; set; }
        public int Cost { get; set; }
        public int SightRange { get; set; }
        public int ResourcePerHealth { get; set; }
        public Model.BuildProperties? Build { get; set; }
        public Model.AttackProperties? Attack { get; set; }
        public Model.RepairProperties? Repair { get; set; }
        public EntityProperties(int size, int buildScore, int destroyScore, bool canMove, int populationProvide, int populationUse, int maxHealth, int cost, int sightRange, int resourcePerHealth, Model.BuildProperties? build, Model.AttackProperties? attack, Model.RepairProperties? repair)
        {
            this.Size = size;
            this.BuildScore = buildScore;
            this.DestroyScore = destroyScore;
            this.CanMove = canMove;
            this.PopulationProvide = populationProvide;
            this.PopulationUse = populationUse;
            this.MaxHealth = maxHealth;
            this.Cost = cost;
            this.SightRange = sightRange;
            this.ResourcePerHealth = resourcePerHealth;
            this.Build = build;
            this.Attack = attack;
            this.Repair = repair;
        }
        public static EntityProperties ReadFrom(System.IO.BinaryReader reader)
        {
            var result = new EntityProperties();
            result.Size = reader.ReadInt32();
            result.BuildScore = reader.ReadInt32();
            result.DestroyScore = reader.ReadInt32();
            result.CanMove = reader.ReadBoolean();
            result.PopulationProvide = reader.ReadInt32();
            result.PopulationUse = reader.ReadInt32();
            result.MaxHealth = reader.ReadInt32();
            result.Cost = reader.ReadInt32();
            result.SightRange = reader.ReadInt32();
            result.ResourcePerHealth = reader.ReadInt32();
            if (reader.ReadBoolean())
            {
                result.Build = Model.BuildProperties.ReadFrom(reader);
            } else
            {
                result.Build = null;
            }
            if (reader.ReadBoolean())
            {
                result.Attack = Model.AttackProperties.ReadFrom(reader);
            } else
            {
                result.Attack = null;
            }
            if (reader.ReadBoolean())
            {
                result.Repair = Model.RepairProperties.ReadFrom(reader);
            } else
            {
                result.Repair = null;
            }
            return result;
        }
        public void WriteTo(System.IO.BinaryWriter writer)
        {
            writer.Write(Size);
            writer.Write(BuildScore);
            writer.Write(DestroyScore);
            writer.Write(CanMove);
            writer.Write(PopulationProvide);
            writer.Write(PopulationUse);
            writer.Write(MaxHealth);
            writer.Write(Cost);
            writer.Write(SightRange);
            writer.Write(ResourcePerHealth);
            if (!Build.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                Build.Value.WriteTo(writer);
            }
            if (!Attack.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                Attack.Value.WriteTo(writer);
            }
            if (!Repair.HasValue)
            {
                writer.Write(false);
            } else
            {
                writer.Write(true);
                Repair.Value.WriteTo(writer);
            }
        }
    }
}
